### PR TITLE
Feature/enable entity id interest

### PR DESCRIPTION
**Pull request by m-samiec**
REVIEW IN COMMIT ORDER

Full integration of USI, as well as a range of small functionality improvements that were made along the way.

Blueprint changes are all minimal config changes.

GDK - https://github.com/spatialos/UnrealGDK/pull/3339

### CODE REVIEW

File: Game/Source/GDKTestGyms/Private/UserExperienceReporter.cpp [hunk 1]
```diff
@@ -8,7 +8,9 @@
 #include "Net/UnrealNetwork.h"
 #include "EngineClasses/SpatialNetDriver.h"
 #include "UserExperienceComponent.h"
+#include "Interop/Connection/SpatialWorkerConnection.h"
 #include "Utils/SpatialMetrics.h"
+#include "Utils/SpatialStatics.h"
```
Comment (MatthewSandfordImprobable) [inline Game/Source/GDKTestGyms/Private/UserExperienceReporter.cpp:13]: What is SpatialStatics.h used for?

File: Game/Source/GDKTestGyms/Private/UserExperienceComponent.cpp [hunk 2]
```diff
@@ -32,10 +32,11 @@ void UUserExperienceComponent::StartRoundtrip()
 {
 	if (OpenRPCs.Contains(RequestKey))
 	{
+		UE_LOG(LogTemp, Warning, TEXT("UserExperience failure: Previous round trip didn't complete in time (%d)."), RequestKey);
```
Comment (MatthewSandfordImprobable) [inline Game/Source/GDKTestGyms/Private/UserExperienceComponent.cpp:35]: If this is a "failure", will we fail NFRs based on the presence of this log?

File: Game/Source/GDKTestGyms/Private/UserExperienceComponent.cpp [hunk 3]
```diff
@@ -96,6 +100,16 @@ float UUserExperienceComponent::CalculateAverageUpdateTimeDelta() const
 	return Avg;
 }
 
+void UUserExperienceComponent::CheckPosition()
+{
+	FVector CurrPos = GetOwner()->GetActorLocation();
+	if (PreviousPos == CurrPos)
+	{
+		UE_LOG(LogTemp, Warning, TEXT("UserExperience failure: Client position not updating (%s). Vel %s"), *PreviousPos.ToString(), *GetOwner()->GetVelocity().ToString());
```
Comment (MatthewSandfordImprobable) [inline Game/Source/GDKTestGyms/Private/UserExperienceComponent.cpp:108]: Similarly, will we fail NFRs based on the presence of this log?

File: Game/Source/GDKTestGyms/Private/LatencyTracer.cpp [hunk 4]
```diff
@@ -3,7 +3,13 @@
 #include "LatencyTracer.h"
 
 #include "Engine/World.h"
+
+#include "EngineClasses/SpatialNetDriver.h"
```
Comment (MatthewSandfordImprobable) [inline Game/Source/GDKTestGyms/Private/LatencyTracer.cpp:7]: Thank you!

File: Game/Source/GDKTestGyms/Public/TestGymsCharacterMovementComp.h [hunk 5]
```diff
@@ -23,7 +23,7 @@ class GDKTESTGYMS_API UTestGymsCharacterMovementComp : public UCharacterMovement
 	void ClientAuthServerMove_Implementation(const FVector_NetQuantize100& ClientLocation, const FVector_NetQuantize10& ClientVelocity, const uint32 PackedPitchYaw);
 	
 	bool bClientAuthMovement = false;
-	int32 ClientMoveFrequency = 10; // Send client position updates to server at this frequency
+	int32 ClientMoveFrequency = 30; // Send client position updates to server at this frequency
```
Comment (MatthewSandfordImprobable) [inline Game/Source/GDKTestGyms/Public/TestGymsCharacterMovementComp.h:26]: Guess this affects all many of our tests. What was the purpose here?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 6]
```diff
@@ -899,6 +890,40 @@ void ABenchmarkGymGameModeBase::ReportAuthoritativePlayerMovement_Implementation
...
+
+	if (!WorkerID.IsEmpty())
+	{
+		LatestClientUXMap.Emplace(WorkerID, UX{RTTime, UpdateTime});
+	}
+
+	AveragedClientRTTMS = 0.f;
+	AveragedClientUpdateTimeDeltaMS = 0.f;
+
+	for (auto& Entry : LatestClientUXMap)
```
Comment (MatthewSandfordImprobable) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:914]: const auto& Entry

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 7]
```diff
@@ -899,6 +890,40 @@ void ABenchmarkGymGameModeBase::ReportAuthoritativePlayerMovement_Implementation
...
+	{
+		AveragedClientRTTMS = FMath::Max(AveragedClientRTTMS, Entry.Value.RTT);
```
Comment (MatthewSandfordImprobable) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:916]: I am probably missing something here but... AveragedClientRTTMS seems to be set to the maximum of all workers ClientRTTMS instead of the average?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 8]
```diff
@@ -147,6 +147,15 @@ void ABenchmarkGymGameModeBase::BeginPlay()
 	{
 		OutputPlayerDensity();
 	}
+
+	if (bLongFormScenario)
+	{
+		// Extend timers to handle longer expected deployment lifetime (required for current long form disco performance test)
+		RequiredPlayerCheckTimer.SetTimer(17 * 60);
+		DeploymentValidTimer.SetTimer(38 * 60);
+		UNFRConstants* NFRConstants = const_cast<UNFRConstants*>(UNFRConstants::Get(GetWorld()));
```
Comment (MatthewSandfordImprobable) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:156]: Feel as if there could be room for a mini rework of how UNFRConstants work but... maybe in future.

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 9]
```diff
@@ -488,6 +488,7 @@ void ABenchmarkGymGameMode::GenerateTestScenarioLocations()
 {
 	const APawn* Pawn = GetDefault<APawn>(SimulatedPawnClass);
 	const float RoamRadius = FMath::Sqrt(Pawn->NetCullDistanceSquared) / 2.0f;
+	const float MinDistanceSq = 500.f * 500.f;	// Move To threshold used in behaviour tree
```
Comment (MatthewSandfordImprobable) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:491]: Nice catch!

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 10]
```diff
@@ -496,6 +497,10 @@ void ABenchmarkGymGameMode::GenerateTestScenarioLocations()
 			FVector PointA = PlayerStream.VRand()*RoamRadius;
 			FVector PointB = PlayerStream.VRand()*RoamRadius;
 			PointA.Z = PointB.Z = 0.0f;
+			if (FVector::DistSquared2D(PointA, PointB) < MinDistanceSq)
+			{
+				PointA *= -1.f;
```
Comment (MatthewSandfordImprobable) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:502]: Does this have the possibility of causing characters to move out of the "density regions"?
Perhaps we just generate random points till we find a valid pair?

**Review by MatthewSandfordImprobable — COMMENTED** [review]
Would you be able to give a quick sum up of what happened in the blueprints? Looks like the component was broken before these changes? 

image

File: Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp [hunk 11]
```diff
@@ -749,6 +880,35 @@ void UTestGymsReplicationGraphNode_GlobalViewTarget::GatherActorListsForConnecti
 	Params.OutGatheredReplicationLists.AddReplicationActorList(ReplicationActorList);
 }
 
+void UTestGymsReplicationGraphNode_GlobalViewTarget::GatherClientInterestedActors(const FConnectionGatherActorListParameters& Params)
+{
+	ReplicationActorList.Reset();
+
+	for (const FNetViewer& CurViewer : Params.Viewers)
+	{
+		if (CurViewer.InViewer)
```
Comment (MatthewSandfordImprobable) [inline Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp:889]: Some explicit checks here

File: Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp [hunk 12]
```diff
@@ -757,6 +917,87 @@ void UTestGymsReplicationGraphNode_GlobalViewTarget::LogNode(FReplicationGraphDe
 	DebugInfo.PopIndent();
 }
 
+
+void UTestGymsReplicationGraphNode_NearestActors::NotifyAddNetworkActor(const FNewReplicatedActorInfo& ActorInfo)
+{
+	ReplicationActorList.Add(ActorInfo.Actor);
+}
+
+bool UTestGymsReplicationGraphNode_NearestActors::NotifyRemoveNetworkActor(const FNewReplicatedActorInfo& ActorInfo, bool bWarnIfNotFound /*= true*/)
```
Comment (MatthewSandfordImprobable) [inline Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp:926]: Maybe we should warn if not found actor to remove?

File: Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp [hunk 13]
```diff
@@ -808,3 +1049,22 @@ FAutoConsoleCommandWithWorldAndArgs ChangeFrequencyBucketsCmd(TEXT("TestGymsRepG
 		Node->SetNonStreamingCollectionSize(Buckets);
 	}
 }));
+
+FAutoConsoleCommandWithWorldAndArgs ChangeDensityCmd(TEXT("TestGymsRepGraph.AlterNearestN"), TEXT("Alters nearest actor density"),
```
Comment (MatthewSandfordImprobable) [inline Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp:1053]: Coool!

File: Game/Source/GDKTestGyms/Private/UserExperienceReporter.cpp (hunk 1 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/Private/UserExperienceReporter.cpp:13]: Removed

File: Game/Source/GDKTestGyms/Private/UserExperienceComponent.cpp (hunk 2 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/Private/UserExperienceComponent.cpp:35]: No it won't, as the client logs aren't parsed. But this log would most likely indicate that the UX metric has failed, and having "UserExperience failure" is useful for bulk searching.

File: Game/Source/GDKTestGyms/Private/UserExperienceComponent.cpp (hunk 3 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/Private/UserExperienceComponent.cpp:108]: Same as above

File: Game/Source/GDKTestGyms/Public/TestGymsCharacterMovementComp.h (hunk 5 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/Public/TestGymsCharacterMovementComp.h:26]: This is only used for client-auth movement rpcs, which are only used in the new disco scenario.

Comment (m-samiec) [reply Game/Source/GDKTestGyms/Public/TestGymsCharacterMovementComp.h:26]: And purpose was to make the scenario more "representative"

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 14]
```diff
@@ -614,7 +614,7 @@ AActor* ABenchmarkGymGameMode::FindPlayerStart_Implementation(AController* Playe
 {
 	if (SpawnManager->GetNumSpawnPoints() == 0)
 	{
-		return Super::FindPlayerStart_Implementation(Player, IncomingName);
+		return nullptr;
```
[the remark this answers is not in the document]
Comment (m-samiec) [reply Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:617]: Yes, as the SpawnManager is only initialised on the auth game mode, if the spatial spawner attempts to access a local non-auth game mode, it should attempt to pass the message to the auth game mode. Previously when running in editor this was calling the super function, which was returning the WorldSettings obj which was used as a starting location, and the Spawn Point logic wasn't respected.

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 7 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:916]: Yeah I'll agree it's confusing. So each server worker averages it's client worker ux metrics, and sends that as a single value to the auth game mode. The auth game mode then looks for the worst aggregated average, from any one workers perspective, and if it's worse than the threshold, fails the test. I'll update the variable naming to try to make that clearer.

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 8 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:156]: Definitely agree

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp (hunk 10 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:502]: PointA and B are relative vectors of the same length (RoamRadius) that are added onto an origin position to generate the two waypoint coordinates. If we generated infinite points and plotted them, we'd get a perfect circle. This is just checking that if the two points that we've generated onto that circle, are too close to one another, flip the position of the first point (so that it's on the other side of the circle).

File: Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp (hunk 12 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp:926]: I don't think that's necessary, we return if the remove was successful or not, so the user of the node could warn if necessary.

**Comment by m-samiec** [comment]
> Would you be able to give a quick sum up of what happened in the blueprints? Looks like the ReplicatedComponent was broken before these changes?
> 
> Also what was the counter object added to the game state?
> 
> image

Mentioned in the PR description that all blueprint changes are just config changes.

ReplicatedComponent_BP was just an option to replicate less data (this isn't enabled by default, but was useful in the stress testing experimentation). As all the characters+npcs use this component, it required re-saving them.

The DiscoGameState involved just added a CounterComponent similar to what's already setup on the BenchmarkGameState_BP class. This is just for outputting the at some frequency the count of specific actor counts into each worker's logs.

File: Game/Source/GDKTestGyms/Private/UserExperienceComponent.cpp [hunk 15]
```diff
@@ -76,10 +77,13 @@ void UUserExperienceComponent::OnRep_ClientTimeTicks(int64 OldTicks)
...
 	Super::OnClientOwnershipGained();
+
 	FTimerHandle Timer;
-	FTimerDelegate Delegate;
-	Delegate.BindUObject(this, &UUserExperienceComponent::StartRoundtrip);
-	GetWorld()->GetTimerManager().SetTimer(Timer, Delegate, 1.0f, true);
+	GetWorld()->GetTimerManager().SetTimer(Timer, this, &UUserExperienceComponent::StartRoundtrip, 1.0f, true);
+
+	FTimerHandle PositionCheckTimer;
+	GetWorld()->GetTimerManager().SetTimer(PositionCheckTimer, this, &UUserExperienceComponent::CheckPosition, 10.0f, true, 10.0f);
```
Comment (simonsarginson) [inline Game/Source/GDKTestGyms/Private/UserExperienceComponent.cpp:85]: Is this safe, will this blow up if `UUserExperienceComponent` get gced from under it?

File: Game/Source/GDKTestGyms/Private/UserExperienceReporter.cpp [hunk 16]
```diff
@@ -96,6 +98,10 @@ void UUserExperienceReporter::ReportMetrics()
 			if (Constants->ClientFPSMetricDelay.HasTimerGoneOff() && GameInstance->GetAveragedFPS() < Constants->GetMinClientFPS())
 			{
 				bFrameRateValid = false;
+				if (const USpatialNetDriver* NetDriver = Cast<USpatialNetDriver>(GameInstance->GetWorld()->GetNetDriver()))
+				{
+					UE_LOG(LogTemp, Error, TEXT("Client %s failed with fps %f below min threshold %f"), *NetDriver->Connection->GetWorkerId(), GameInstance->GetAveragedFPS(), Constants->GetMinClientFPS());
```
Comment (simonsarginson) [inline Game/Source/GDKTestGyms/Private/UserExperienceReporter.cpp]: `LogTemp` ?

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp (hunk 10 above)
Comment (simonsarginson) [reply Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:502]: Worth a comment?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 17]
```diff
@@ -730,6 +751,16 @@ void ABenchmarkGymGameModeBase::ReadWorkerFlagValues(USpatialWorkerFlags* Spatia
 		SetLifetime(FCString::Atoi(*LifetimeString));
 	}
 
+	if (SpatialWorkerFlags->GetWorkerFlag(CubeRespawnBaseTimeWorkerFlag, CubeRespawnBaseTimeString))
+	{
+		CubeRespawnBaseTime = FCString::Atof(*CubeRespawnBaseTimeString);
```
Comment (simonsarginson) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:756]: This returns 0.0 on failure to make any sense of the string, is that acceptable?
(It calls `_tstof`  -> `_wtof` under the hood https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/atof-atof-l-wtof-wtof-l?view=msvc-170).

Comment (simonsarginson) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:756]: Would actually suggest wrapping this in 
```
if (CubeRespawnBaseTimeString.IsNumeric())
{
}
// or
checkf(CubeRespawnBaseTimeString.IsNumeric(), TEXT("Not a valid number"));
```

Comment (simonsarginson) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:756]: Might even be worth wrapping into a free function since there is a lot of string to float in this commit.

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (simonsarginson) [reply Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:617]: worth a comment?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 18]
```diff
@@ -149,6 +149,14 @@ void ABenchmarkGymGameModeBase::BeginPlay()
 	}
 }
 
+void ABenchmarkGymGameModeBase::OnAuthorityLost()
+{
+	// ABenchmarkGymGameModeBase doesn't support authority transfer
```
Comment (simonsarginson) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:163]: Comment can be wrapped into a `ensuref` ?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 19]
```diff
@@ -419,8 +427,7 @@ void ABenchmarkGymGameModeBase::Tick(float DeltaSeconds)
 	
 	// PrintMetricsTimer needs to be reset at the the end of ABenchmarkGymGameModeBase::Tick.
 	// This is so that the above function have a chance to run logic dependant on PrintMetricsTimer.HasTimerGoneOff().
-	if (HasAuthority() &&
-		PrintMetricsTimer.HasTimerGoneOff())
+	if (PrintMetricsTimer.HasTimerGoneOff())
```
Comment (simonsarginson) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:439]: Why this change?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 20]
```diff
@@ -99,11 +100,10 @@ ABenchmarkGymGameModeBase::ABenchmarkGymGameModeBase()
 	, UXAuthActorCount(0)
 	, PrintMetricsTimer(10)
 	, TestLifetimeTimer(0)
-	, TickActorCountTimer(60) // 1-minutes to allow workers to get setup and the deployment to get into a stable state
 	, TimeSinceLastCheckedTotalActorCounts(0.0f)
 	, bHasRequiredPlayersCheckFailed(false)
-	, RequiredPlayerCheckTimer(11*60) // 1-minute later then RequiredPlayerReportTimer to make sure all the workers had reported their migration
-	, DeploymentValidTimer(16*60) // 16-minute window to check between
+	, RequiredPlayerCheckTimer(11*60) // all clients should have joined by this point
```
Comment (simonsarginson) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:105]: seconds / minutes? worth calling out?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 8 above)
Comment (simonsarginson) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:156]: +1

File: Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp [hunk 21]
```diff
@@ -481,6 +560,45 @@ void UTestGymsReplicationGraphNode_AlwaysRelevant_ForConnection::ResetGameWorldS
 	AlwaysRelevantStreamingLevelsNeedingReplication.Empty();
 }
 
+void UTestGymsReplicationGraphNode_AlwaysRelevant_ForConnection::GatherClientInterestedActors(const FConnectionGatherActorListParameters& Params)
+{
+	InterestedActorList.Reset();
```
Comment (simonsarginson) [inline Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp:565]: `InterestedActorList.Reset(Params.Viewers.Num()*3);`
Take a greedy guess at space requirements and reserve?

File: Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp [hunk 22]
```diff
@@ -481,6 +560,45 @@ void UTestGymsReplicationGraphNode_AlwaysRelevant_ForConnection::ResetGameWorldS
...
+			}
+		}
+	}
+
+	Params.OutGatheredReplicationLists.AddReplicationActorList(InterestedActorList);
+
+	UTestGymsReplicationGraph* TestGymsGraph = CastChecked<UTestGymsReplicationGraph>(GetOuter());
+	TMap<FName, FActorRepListRefView>& AlwaysRelevantStreamingLevelActors = TestGymsGraph->AlwaysRelevantStreamingLevelActors;
+
+	for (int32 Idx = AlwaysRelevantStreamingLevelsNeedingReplication.Num() - 1; Idx >= 0; --Idx)
```
Comment (simonsarginson) [inline Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp:584]: Why backwards?

File: Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp [hunk 23]
```diff
@@ -749,6 +880,35 @@ void UTestGymsReplicationGraphNode_GlobalViewTarget::GatherActorListsForConnecti
 	Params.OutGatheredReplicationLists.AddReplicationActorList(ReplicationActorList);
 }
 
+void UTestGymsReplicationGraphNode_GlobalViewTarget::GatherClientInterestedActors(const FConnectionGatherActorListParameters& Params)
+{
+	ReplicationActorList.Reset();
```
Comment (simonsarginson) [inline Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp:885]: reserve size?

File: Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp [hunk 24]
```diff
@@ -757,6 +917,87 @@ void UTestGymsReplicationGraphNode_GlobalViewTarget::LogNode(FReplicationGraphDe
...
+void UTestGymsReplicationGraphNode_NearestActors::NotifyAddNetworkActor(const FNewReplicatedActorInfo& ActorInfo)
+{
+	ReplicationActorList.Add(ActorInfo.Actor);
+}
+
+bool UTestGymsReplicationGraphNode_NearestActors::NotifyRemoveNetworkActor(const FNewReplicatedActorInfo& ActorInfo, bool bWarnIfNotFound /*= true*/)
+{
+	bool bRemovedSomething = false;
+	bRemovedSomething = ReplicationActorList.RemoveFast(ActorInfo.Actor);
+	return bRemovedSomething;
```
Comment (simonsarginson) [inline Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp:930]:
```
	return ReplicationActorList.RemoveFast(ActorInfo.Actor);
```
Could just be this?

File: Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp [hunk 25]
```diff
@@ -757,6 +917,87 @@ void UTestGymsReplicationGraphNode_GlobalViewTarget::LogNode(FReplicationGraphDe
...
+
+void UTestGymsReplicationGraphNode_NearestActors::GatherActorListsForConnection(const FConnectionGatherActorListParameters& Params)
+{
+	// Return all actors for replication
+	if (ReplicationActorList.Num() > 0)
+	{
+		FGlobalActorReplicationInfoMap& GlobalMap = *GraphGlobals->GlobalActorReplicationInfoMap;
+
+		// Cache actor location
+		for (FActorRepListType& Actor : ReplicationActorList)
```
Comment (simonsarginson) [inline Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp:941]: const?

File: Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp [hunk 26]
```diff
@@ -757,6 +917,87 @@ void UTestGymsReplicationGraphNode_GlobalViewTarget::LogNode(FReplicationGraphDe
...
+		SortedActors.Reset(ActorCount);
+
+		ensure(Params.Viewers.Num() == 1);	// Don't support multiple viewers for interest calculation
+		const FNetViewer& Viewer = Params.Viewers[0];
+
+		for (AActor* Actor : ReplicationActorList)
+		{
+			FGlobalActorReplicationInfo& ActorRepInfo = GlobalMap.Get(Actor);
+
+			float DistanceToViewer = (Viewer.ViewLocation - ActorRepInfo.WorldLocation).SizeSquared();	// check for max size?
```
Comment (simonsarginson) [inline Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp:970]: const?

File: Game/Source/GDKTestGyms/Private/UserExperienceReporter.cpp [hunk 27]
```diff
@@ -100,7 +100,7 @@ void UUserExperienceReporter::ReportMetrics()
 				bFrameRateValid = false;
 				if (const USpatialNetDriver* NetDriver = Cast<USpatialNetDriver>(GameInstance->GetWorld()->GetNetDriver()))
 				{
-					UE_LOG(LogTemp, Error, TEXT("Client %s failed with fps %f below min threshold %f"), *NetDriver->Connection->GetWorkerId(), GameInstance->GetAveragedFPS(), Constants->GetMinClientFPS());
```
Comment (simonsarginson) [inline Game/Source/GDKTestGyms/Private/UserExperienceReporter.cpp]: LogTemp?

File: Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp [hunk 28]
```diff
@@ -855,12 +855,12 @@ void UTestGymsReplicationGraphNode_GlobalViewTarget::GatherActorListsForConnecti
 
 	for (const FNetViewer& CurViewer : Params.Viewers)
 	{
-		if (CurViewer.InViewer)
+		if (CurViewer.InViewer != nullptr)
```
Comment (simonsarginson) [inline Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp:858]:
```
if (IsValid(CurViewer.InViewer))
```

File: Game/Source/GDKTestGyms/Private/UserExperienceComponent.cpp (hunk 15 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/Private/UserExperienceComponent.cpp:85]: Good point - will unbind on destroy

File: Game/Source/GDKTestGyms/Private/UserExperienceReporter.cpp (hunk 16 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/Private/UserExperienceReporter.cpp]: Yeah - we use it here and elsewhere in small classes within the test gyms

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 17 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:756]: Yeah it would be invalid to pass it anything but a number. We also output the values set by these calls in a general log at the bottom of this function, so the user could debug if it's a problem.

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 19 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:439]: This runs on non-auth servers now. Part of the "allow UX reporting to work in zoning scenarios commit".

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 17 above)
Comment (simonsarginson) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:756]: But it won't be an error, could be difficult to spot if we are just leaning on printing output values?

File: Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp (hunk 21 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp:565]: This will typically be the same from frame to frame, and reset will allow the internal array to persist its size, so don't think setting any size is necessary.

File: Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp (hunk 22 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp:584]: Based on existing code in GatherActorListsForConnection.

File: Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp (hunk 23 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp:885]: Will persist size, so doesn't seem necessary.

File: Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp (hunk 28 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp:858]: Have seen that ConditionalAdd actually does validation checks internally, so have removed this check entirely.

File: Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp (hunk 21 above)
Comment (simonsarginson) [reply Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp:565]: it doesn't shrink on reset, fair dues.

File: Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp (hunk 12 above)
Comment (simonsarginson) [reply Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp:914]: Seems a bit awkward to ignore part of the API surface of this virtual function though? I would be pretty surprised if I called this on a base class pointer and it did not warn.

Comment (m-samiec) [reply Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp:915]: Oh you guys are right, I completely missed the `bWarnIfNotFound` variable - will fix.

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 17 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:755]: I think this is fine. If there was an issue with the scenario, first thing you'd check is the settings.

Comment (simonsarginson) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:755]: ok agree to disagree, not a fan of silent errors, you have more context so consider it resolved : ).

**Review by simonsarginson — APPROVED** [review]
LGTM

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 7 above)
Comment (MatthewSandfordImprobable) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:917]: Perhaps AveragedClientRTTMS should be WorstAveragedClientRTTMS

File: Game/Source/GDKTestGyms/Private/UserExperienceReporter.cpp [hunk 29]
```diff
@@ -96,6 +97,10 @@ void UUserExperienceReporter::ReportMetrics()
 			if (Constants->ClientFPSMetricDelay.HasTimerGoneOff() && GameInstance->GetAveragedFPS() < Constants->GetMinClientFPS())
 			{
 				bFrameRateValid = false;
+				if (const USpatialNetDriver* NetDriver = Cast<USpatialNetDriver>(GameInstance->GetWorld()->GetNetDriver()))
```
Comment (MatthewSandfordImprobable) [inline Game/Source/GDKTestGyms/Private/UserExperienceReporter.cpp:100]: Probably want some log even if netdriver is null. Maybe the worker id can just be invalid if null driver.

**Review by MatthewSandfordImprobable — APPROVED** [review]
